Implement plan_mdct and plan_imdct

Ashymad opened the pull request:
These two functions return specialized function for computing MDCT (or IMDCT) on an input array of known length (similar to `FFTW.plan_dct`). I based the implementation on [this paper](https://pdfs.semanticscholar.org/2f26/a658836927331d559e723ac36b8dab911b14.pdf). Using the pre-planned functions is around 10 times faster on my PC than the not pre-planned versions.

Ashymad committed:
Implement plan_mdct and plan_imdct

stevengj commented:
This looks exactly equivalent to the `mdct` algorithm that is already implemented, except that it pre-computes the FFTW plan and does the ± preprocessing via sparse-matrix operations.  The latter is almost certainly slower than the loop that we already have, and requires allocation of additional arrays because it is not in-place.

So if you want a `plan_mdct` version, I would recommend basing it on the current code instead and just precomputing the plan.

Ashymad commented:
I suppose you mean something like this?
```Julia
function plan_mdct_sj(X::AbstractVector{T}) where {T<:Number}
    sz = length(X)
    if isodd(sz)
        throw(ArgumentError("mdct requires an even-length vector"))
    end
    N = div(sz, 2);
    if isodd(N)
        throw(ArgumentError("mdct requires a multiple-of-4 vector length"))
        # FIXME: handle odd case via DCT-III?
    end
    N2 = div(N,2)
    r2r_plan = plan_r2r!(X[1:N], REDFT11)
    return function(X::AbstractVector{T}) where {T<:Number}
        Y = fftwsimilar(X, N)
        for i = 1:N2
            Y[i] = -0.5 * (X[(3*N2+1)-i] + X[3*N2+i])
            Y[N2+i] = -0.5 * (X[(2*N2+1)-i] - X[i])
        end
        return r2r_plan*Y
    end
end
```
If I benchmark this...
```Julia
using BenchmarkTools
using MDCT

tst = rand(1000)

mdct_plan = plan_mdct(tst)
mdct_sj_plan = MDCT.plan_mdct_sj(tst)
```
```Julia
julia> @btime mdct_plan(tst);
5.708 μs (1 allocation: 4.06 KiB)
julia> @btime mdct_sj_plan(tst);
14.281 μs (503 allocations: 12.00 KiB)
```

Still, my code is ~3x faster and requires ~500x less allocations

stevengj commented:
The huge number of allocations is a clue that there is an inference problem in your code.  In this case, you are running into JuliaLang/julia#15276.   If I instead define the variables `N`, `N2`, and `r2r_plan` in a `let N = div(sz, 2), N2 = div(N,2), r2r_plan = plan_r2r!(X[1:N], REDFT11)` statement, then I get:
```jl
julia> @btime mdct_sj_plan(tst);
  4.711 μs (3 allocations: 4.19 KiB)
```

Also, I think we want to follow FFTW.jl and AbstractFFTs.jl in defining an `MDCT.Plan` data structure instead of using an anonymous function.  That is, define something like:
```jl
struct Plan{T<:fftwNumber} <: AbstractFFTs.Plan{T}
    plan:: r2rFFTWPlan # plan_r2r! REDFT11 plan
    N2::Int
    pinv::ScaledPlan
end
Base.size(p::Plan) = (p.N2, 2*p.N2)
function mul!(Y::StridedArray{T}, p:Plan{T}, X::AbstractArray{T})
    N2 = p.N2
    @boundscheck (length(X) == 2N2 && length(Y) == N2) || throw(DimensionMismatch())
    @inbounds for i = 1:N2
        Y[i] = -0.5 * (X[(3*N2+1)-i] + X[3*N2+i])
        Y[N2+i] = -0.5 * (X[(2*N2+1)-i] - X[i])
    end
    return p.plan * Y
end
```
and so on.

Ashymad commented:
Wow I didn't know about this issue, This makes closures much more annoying to deal with... Anyway thank you very much for your help, I'll work on this after the weekend :)

Ashymad commented:
I've reimplemented the functions with a Plan structure and using the original code. Could you take a look?

stevengj commented on src/MDCT.jl:
```diff
@@ -63,4 +67,60 @@ end
 imdct(X::AbstractVector{T}) where {T<:Number} =
     imdct(copy!(fftwsimilar(X, size(X)), X))
 
+mutable struct MDCTPlan{T<:fftwNumber, N, inv} <: AbstractFFTs.Plan{T}
```
Can just be called `Plan` since it is already in the `MDCT` namespace.

Ashymad replied:
Yeah, if You prefer. I named it in the same way FFTW.FFTWPlan is named.

stevengj commented on src/MDCT.jl:
```diff
@@ -63,4 +67,60 @@ end
+        # FIXME: handle odd case via DCT-III?
+    end
+    N = div(sz, 2);
+    return MDCTPlan{T, N, false}(plan_r2r!(view(X, 1:N), REDFT11))
```
Probably `fftwsimilar(X, N)` rather than a `view`.

stevengj commented on src/MDCT.jl:
```diff
@@ -63,4 +67,60 @@ end
+    return mul!(Y, p, X)
+end
+
+function plan_mdct(X::AbstractVector{T}) where {T<:Number}
```
Should take `flags=FFTW.ESTIMATE, timelimit=Inf` keyword arguments.

Ashymad replied:
Done

Ashymad committed:
Make requested changes; change Array to Vector

stevengj commented on src/MDCT.jl:
```diff
@@ -63,4 +67,60 @@ end
+
+function mul!(Z::StridedVector{T}, p::Plan{T, N, true}, X::AbstractVector{T}) where {T<:Number, N}
+    @boundscheck length(Z) == N || throw(DimensionMismatch())
+    Y = p.plan*copy(X)
```
Seems like we should create an out-of-place plan for this case, so that we can just do `Y = p.plan * X`.

stevengj replied:
(i.e. `plan_r2r`, not `plan_r2r!`)

Ashymad replied:
https://github.com/stevengj/MDCT.jl/blob/75d4081608dc8ee2fc6610231286e5a67f85824a/src/MDCT.jl#L45
https://github.com/stevengj/MDCT.jl/blob/75d4081608dc8ee2fc6610231286e5a67f85824a/src/MDCT.jl#L63-L64
Now that I look a this I am not sure what to do... Why is `fftwsimilar` needed in the case of `AbstractVector{T}` but not needed when the argument is an `StridedVector{T}`? In both cases `T<:Number`. Also why fftwsimilar doesn't return a `StridedArray{T}`? If this is indeed necessary then maybe it would be better to just change `copy()` to `copy!(fftwsimilar())`, so we don't have to keep both in-place and regular plans in MDC.Plan (although this would decrease performance, because using in-place plan on a copy is less performant then using a regular plan)

stevengj replied:
An `Array` is a subtype of `StridedArray`.

stevengj commented:
Also needs documentation in the README.

Ashymad committed:
Use regular plan for imdct

Ashymad committed:
Add desxription in README

Ashymad commented:
I've added documentation and changed the plan for the imdct so it's not in-place, although I'm still not sure If we should make a special case for argument of type `AbstractArray` like in the regular `imdct` function

stevengj commented:
Looks good, thanks!
